adding basic support for OpenSSH certificate signing

rogerhu opened the pull request:
OpenSSH has support for `CertificateFile` (https://man7.org/linux/man-pages/man5/ssh_config.5.html) and PKCS11Provider (https://stackoverflow.com/a/52139555).  I've introduced two extra config values, `ssh_bastion_identity_certificate_file` and `ssh_identity_certificate_file` that can be specified.  Because the certificate file can expire, I've also added verification checks to flag if the file has expired or not.

The Golang SSH client needs to be modified to add an extra CertSigner.  

Tests for the basic flow included.  Without a bastion host that requires this, however, I can only be manually tested.  Or is there a better way?

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=9521)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=9521)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=9521).</sub>

rogerhu commented on communicator/ssh/connect_test.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+// Rename to Uppercase to test and change host constant
+func testBastionConnectFunc(t *testing.T) {
```
Not sure -- how to test this without having things setup properly in an SSH environment.  The lint checks are failing.  Should I just remove this file?

SwampDragons replied:
Keep the file for now; but we can put it behind an acceptance test flag: 
```
	if os.Getenv("PACKER_ACC") == "" {
		t.Skip("This test is only run with PACKER_ACC=1")
	}
```

rogerhu replied:
Fixed!  Thanks for the pointer.

SwampDragons commented on helper/communicator/config.go:
```diff
@@ -480,9 +507,20 @@ func (c *Config) prepareSSH(ctx *interpolate.Context) []error {
-				errs = append(errs, fmt.Errorf(
-					"ssh_bastion_private_key_file is invalid: %s", err))
+			} else {
+				var certPath string
```
same nitpick as above.

rogerhu replied:
Done!

SwampDragons commented on helper/ssh/ssh.go:
```diff
@@ -42,3 +51,62 @@ func FileSigner(path string) (ssh.Signer, error) {
+	}
+
+	// Load private key
+	if err != nil {
```
Where's this error coming from? Looks like you may have missed a step or copypasted the error check twice.

rogerhu replied:
Yeah I fixed - thanks for the catch.

SwampDragons commented on helper/ssh/ssh.go:
```diff
@@ -42,3 +51,62 @@ func FileSigner(path string) (ssh.Signer, error) {
+		return nil, fmt.Errorf("unable to parse public key: %v", err)
+	}
+
+	certificate := pk.(*ssh.Certificate)
```
If this assertion fails, Packer will crash. Let's do it as a check instead: 
```suggestion
	certificate, ok := pk.(*ssh.Certificate)
	if !ok {
       	return nil, fmt.Errorf("Error loading certificate")
	}
```

rogerhu replied:
Fixed

rogerhu commented on helper/communicator/config.go:
```diff
@@ -462,9 +477,21 @@ func (c *Config) prepareSSH(ctx *interpolate.Context) []error {
-			errs = append(errs, fmt.Errorf(
-				"ssh_private_key_file is invalid: %s", err))
+		} else {
+			var certPath string
```
Fixed!

rogerhu commented:
> This looks like really great start -- So far I just have a couple of minor suggestions and a question inline.
> 
> You're right that this isn't totally finished yet:
> 
> > The Golang SSH client needs to be modified to add an extra CertSigner.
> 
> I believe that'll need to happen in config.go in the SSHConfigFunc(). Until we add the CertSigner to that Auth field on the client, this won't have any effect for non-bastion certs. Looks like you've got the bastion certs all set up already though.

Yep I have bastion certs.  I've added it to SSHConfigFunc() with similar functionality.  What's strange is this code aggregates all the private keys before using the signer.  What's the reason?  I have to assume every private key needs to be signed.

SwampDragons commented:
You can only provide one "publickey" auth method to the ssh client config, and that signer loops over all the keys and signs them as it tries them, I think. If you provide multiple publickey signers, I believe it'll only use the first.

rogerhu commented:
I fixed it so that it reads the certificate file once and now just signs every private key - hope that makes it more clear?

Not sure what's up with the test failure -- seems unrelated to these changes.  I rebased to upstream to make sure I have the latest changes.

SwampDragons commented:
those test failures are related to the ssh config, which you're changing -- I believe they're relevant to this changeset.

rogerhu commented:
Ok!  Finally figured out the failing test.  Take a look again?

SwampDragons commented:
Awesome -- not gonna have time to get to this tonight, but I'll review it tomorrow.  Thanks!

SwampDragons commented:
I'm rebasing and regenerating the generated files to resolve conflicts. Then I'll merge. This works as expected.

rogerhu commented:
Thank you!
